File: TrainBlink/android/app/src/main/java/com/trainblink/app/models/ChatMessage.kt
package com.trainblink.app.models

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID

/**
 * Chat Message Model
 * Represents a chat message sent between peers
 * Android equivalent of iOS ChatMessage.swift
 */
data class ChatMessage(
    val id: String = UUID.randomUUID().toString(),
    val text: String,
    val senderId: String,
    val receiverId: String,
    val timestamp: Date = Date(),
    var isEncrypted: Boolean = false,
    var isRead: Boolean = false,
    var deliveryStatus: MessageDeliveryStatus = MessageDeliveryStatus.PENDING,
    var deliveredAt: Date? = null,
    var readAt: Date? = null,
    // Ephemeral (Feature 6)
    var isEphemeral: Boolean = false,
    var expiresAt: Date? = null,
    var isExpired: Boolean = false
) {
    companion object {
        /**
         * Create a text message
         */
        fun textMessage(text: String, from: String, to: String): ChatMessage {
            return ChatMessage(
                text = text,
                senderId = from,
                receiverId = to,
                deliveryStatus = MessageDeliveryStatus.PENDING
            )
        }

        /**
         * Create an ephemeral message (auto-deletes after 10 seconds)
         */
        fun ephemeralMessage(
            text: String,
            from: String,
            to: String,
            lifetimeSeconds: Long = 10
        ): ChatMessage {
            val now = Date()
            val expiresAt = Date(now.time + lifetimeSeconds * 1000)

            return ChatMessage(
                text = text,
                senderId = from,
                receiverId = to,
                timestamp = now,
                deliveryStatus = MessageDeliveryStatus.PENDING,
                isEphemeral = true,
                expiresAt = expiresAt
            )
        }
    }

    /**
     * Whether this message was sent by me
     */
    fun isSentByMe(myId: String): Boolean = senderId == myId

    /**
     * Whether this message was received by me
     */
    fun isReceivedByMe(myId: String): Boolean = receiverId == myId

    /**
     * Time ago string (e.g., "2m ago", "1h ago")
     */
    val timeAgoString: String
        get() {
            val interval = (Date().time - timestamp.time) / 1000 // seconds

            return when {
                interval < 60 -> "Just now"
                interval < 3600 -> "${interval / 60}m ago"
                interval < 86400 -> "${interval / 3600}h ago"
                else -> "${interval / 86400}d ago"
            }
        }

    /**
     * Formatted time string
     */
    val formattedTime: String
        get() {
            val formatter = SimpleDateFormat("HH:mm", Locale.getDefault())
            return formatter.format(timestamp)
        }

    /**
     * Message preview (truncated)
     */
    val preview: String
        get() = if (text.length > 50) text.take(50) + "..." else text

    /**
     * Time remaining until message expires (in seconds)
     */
    val timeRemainingSeconds: Long?
        get() {
            if (!isEphemeral || expiresAt == null) return null
            val remaining = (expiresAt.time - Date().time) / 1000
            return if (remaining > 0) remaining else 0
        }

    /**
     * Whether the message should be deleted now
     */
    val shouldDelete: Boolean
        get() {
            if (!isEphemeral) return false
            val expiry = expiresAt ?: return false
            return Date() >= expiry || isExpired
        }

    /**
     * Countdown string for ephemeral messages (e.g., "5s")
     */
    val countdownString: String
        get() {
            val remaining = timeRemainingSeconds ?: return ""
            return "${remaining}s"
        }

    /**
     * Mark message as delivered
     */
    fun markAsDelivered(): ChatMessage {
        return copy(
            deliveryStatus = MessageDeliveryStatus.DELIVERED,
            deliveredAt = Date()
        )
    }

    /**
     * Mark message as read
     */
    fun markAsRead(): ChatMessage {
        return copy(
            isRead = true,
            readAt = Date(),
            deliveryStatus = if (deliveryStatus == MessageDeliveryStatus.DELIVERED) {
                MessageDeliveryStatus.READ
            } else {
                deliveryStatus
            }
        )
    }

    /**
     * Mark message as failed
     */
    fun markAsFailed(): ChatMessage {
        return copy(deliveryStatus = MessageDeliveryStatus.FAILED)
    }

    /**
     * Mark message as expired (for ephemeral messages)
     */
    fun markAsExpired(): ChatMessage {
        return copy(isExpired = true)
    }
}

/**
 * Message Delivery Status
 */
enum class MessageDeliveryStatus {
    PENDING,
    SENDING,
    SENT,
    DELIVERED,
    READ,
    FAILED;

    val iconName: String
        get() = when (this) {
            PENDING -> "schedule"
            SENDING -> "upload"
            SENT -> "check"
            DELIVERED -> "done_all"
            READ -> "done_all" // Could be a different color
            FAILED -> "error"
        }
}

File: TrainBlink/ios/TrainBlink/Models/ChatMessage.swift
//
//  ChatMessage.swift
//  TrainBlink
//
//  Feature 5: 1-on-1 Chat Rooms
//  Model representing a chat message
//
//  Created by TrainBlink Team
//  Copyright © 2025 TrainBlink. All rights reserved.
//

import Foundation

/// Chat message sent between peers
struct ChatMessage: Identifiable, Codable, Hashable, Equatable {

    // MARK: - Properties

    let id: String
    let text: String
    let senderId: String
    let receiverId: String
    let timestamp: Date
    var isEncrypted: Bool
    var isRead: Bool
    var deliveryStatus: MessageDeliveryStatus

    // Metadata
    var encryptedData: Data?          // Encrypted message data
    var encryptionKeyId: String?      // Reference to encryption key
    var deliveredAt: Date?
    var readAt: Date?

    // Ephemeral (Feature 6)
    var isEphemeral: Bool             // Whether this message auto-deletes
    var expiresAt: Date?              // When the message will be deleted
    var isExpired: Bool               // Whether the message has expired

    // MARK: - Initialization

    init(
        id: String = UUID().uuidString,
        text: String,
        senderId: String,
        receiverId: String,
        timestamp: Date = Date(),
        isEncrypted: Bool = false,
        isRead: Bool = false,
        deliveryStatus: MessageDeliveryStatus = .pending,
        isEphemeral: Bool = false,
        expiresAt: Date? = nil
    ) {
        self.id = id
        self.text = text
        self.senderId = senderId
        self.receiverId = receiverId
        self.timestamp = timestamp
        self.isEncrypted = isEncrypted
        self.isRead = isRead
        self.deliveryStatus = deliveryStatus
        self.isEphemeral = isEphemeral
        self.expiresAt = expiresAt
        self.isExpired = false
    }

    // MARK: - Computed Properties

    /// Whether this message was sent by me
    func isSentByMe(myId: String) -> Bool {
        return senderId == myId
    }

    /// Whether this message was received by me
    func isReceivedByMe(myId: String) -> Bool {
        return receiverId == myId
    }

    /// Time ago string (e.g., "2m ago", "1h ago")
    var timeAgoString: String {
        let interval = Date().timeIntervalSince(timestamp)

        if interval < 60 {
            return "Just now"
        } else if interval < 3600 {
            let minutes = Int(interval / 60)
            return "\(minutes)m ago"
        } else if interval < 86400 {
            let hours = Int(interval / 3600)
            return "\(hours)h ago"
        } else {
            let days = Int(interval / 86400)
            return "\(days)d ago"
        }
    }

    /// Formatted time string
    var formattedTime: String {
        let formatter = DateFormatter()
        formatter.timeStyle = .short
        return formatter.string(from: timestamp)
    }

    /// Message preview (truncated)
    var preview: String {
        if text.count > 50 {
            return String(text.prefix(50)) + "..."
        }
        return text
    }

    // MARK: - Ephemeral Message Properties (Feature 6)

    /// Time remaining until message expires (in seconds)
    var timeRemainingSeconds: TimeInterval? {
        guard isEphemeral, let expiresAt = expiresAt else {
            return nil
        }
        let remaining = expiresAt.timeIntervalSince(Date())
        return max(0, remaining)
    }

    /// Whether the message should be deleted now
    var shouldDelete: Bool {
        guard isEphemeral else { return false }
        guard let expiresAt = expiresAt else { return false }
        return Date() >= expiresAt || isExpired
    }

    /// Countdown string for ephemeral messages (e.g., "5s")
    var countdownString: String {
        guard let remaining = timeRemainingSeconds else {
            return ""
        }
        let seconds = Int(ceil(remaining))
        return "\(seconds)s"
    }

    // MARK: - Update Methods

    /// Mark message as delivered
    mutating func markAsDelivered() {
        deliveryStatus = .delivered
        deliveredAt = Date()
    }

    /// Mark message as read
    mutating func markAsRead() {
        isRead = true
        readAt = Date()
        if deliveryStatus == .delivered {
            deliveryStatus = .read
        }
    }

    /// Mark message as failed
    mutating func markAsFailed() {
        deliveryStatus = .failed
    }

    /// Mark message as expired (for ephemeral messages)
    mutating func markAsExpired() {
        isExpired = true
    }

    // MARK: - Static Constructors

    /// Create a text message
    static func textMessage(
        text: String,
        from senderId: String,
        to receiverId: String
    ) -> ChatMessage {
        return ChatMessage(
            text: text,
            senderId: senderId,
            receiverId: receiverId,
            timestamp: Date(),
            deliveryStatus: .pending
        )
    }

    /// Create an ephemeral message (auto-deletes after 10 seconds)
    static func ephemeralMessage(
        text: String,
        from senderId: String,
        to receiverId: String,
        lifetimeSeconds: TimeInterval = 10.0
    ) -> ChatMessage {
        let now = Date()
        let expiresAt = now.addingTimeInterval(lifetimeSeconds)

        return ChatMessage(
            text: text,
            senderId: senderId,
            receiverId: receiverId,
            timestamp: now,
            deliveryStatus: .pending,
            isEphemeral: true,
            expiresAt: expiresAt
        )
    }

    // MARK: - Equatable

    static func == (lhs: ChatMessage, rhs: ChatMessage) -> Bool {
        return lhs.id == rhs.id
    }

    // MARK: - Hashable

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

// MARK: - Supporting Types

/// Message delivery status
enum MessageDeliveryStatus: String, Codable {
    case pending      // Waiting to be sent
    case sending      // Currently sending
    case sent         // Sent successfully
    case delivered    // Delivered to recipient
    case read         // Read by recipient
    case failed       // Failed to send
}

/// Message status icon
extension MessageDeliveryStatus {
    var iconName: String {
        switch self {
        case .pending:
            return "clock"
        case .sending:
            return "arrow.up.circle"
        case .sent:
            return "checkmark"
        case .delivered:
            return "checkmark.circle"
        case .read:
            return "checkmark.circle.fill"
        case .failed:
            return "exclamationmark.circle"
        }
    }
}
